mkosi/__init__.py

```diff
@@ -4170,45 +4162,59 @@ def install_unified_kernel(
-                # kernel image files can be found either in /usr/lib/modules/<kernel-version>/vmlinuz or in
-                # /boot depending on the distro. By invoking the kernel-install script directly, we can pass
-                # the empty string as the kernel image which causes the script to not pass the --kernel-image
-                # option to dracut so it searches the image for us.
-                cmdline = [
-                    "/etc/kernel/install.d/50-mkosi-dracut-unified-kernel.install",
-                    "add",
-                    kver.name,
-                    f"{prefix}/{args.machine_id}/{kver.name}",
-                    "",
-                ]
-                if args.distribution == Distribution.gentoo:
-                    from .gentoo import ARCHITECTURES
-
-                    _, kimg_path = ARCHITECTURES[args.architecture or "x86_64"]
-
-                    cmdline[4] = f"/usr/src/linux-{kver.name}/{kimg_path}"
-
-                # Pass some extra meta-info to the script via
-                # environment variables. The script uses this to name
-                # the unified kernel image file
-                env = {}
-                if args.image_id is not None:
-                    env["IMAGE_ID"] = args.image_id
-                if args.image_version is not None:
-                    env["IMAGE_VERSION"] = args.image_version
-                if root_hash is not None:
-                    env["USRHASH" if args.usr_only else "ROOTHASH"] = root_hash
-
-                run_workspace_command(args, root, cmdline, env=env)
+            if args.image_version:
+                boot_binary = root / prefix / f"EFI/Linux/{image_id}_{args.image_version}.efi"
+            elif root_hash:
+                boot_binary = root / prefix / f"EFI/Linux/{image_id}-{kver}-{root_hash}.efi"
+            else:
+                boot_binary = root / prefix / f"EFI/Linux/{image_id}-{kver}.efi"
+
+            if root.joinpath("etc/kernel/cmdline").exists():
+                boot_options = root.joinpath("etc/kernel/cmdline").read_text().strip()
+            elif root.joinpath("/usr/lib/kernel/cmdline").exists():
+                boot_options = root.joinpath("usr/lib/kernel/cmdline").read_text().strip()
+            else:
+                boot_options = ""
+
+            if root_hash:
+                option = "usrhash" if args.usr_only else "roothash"
+                boot_options = f"{boot_options} {option}={root_hash}"
+            elif partlabel:
+                boot_options = f"{boot_options} root=PARTLABEL={partlabel}"
+
+            osrelease = root / "usr/lib/os-release"
+            # Write to something in $BOOT because everything else is read-only.
+            cmdline = root / prefix / "cmdline"
+            cmdline.write_text(boot_options)
+            initrd = root / prefix / args.machine_id / kver / "initrd"
+
+            cmd: Sequence[PathString] = [
```

Comment:
Could you add a todo comment here or maybe a cast? This is a mypy bug. It correctly identifies the type of `root` and `boot_binary` (and without them also `cmd`), but leaving either in result in `cmd` being wrongly identified it as `list[object]`. A cast might be preferable, since it will be flagged automatically, when in the future the cast is identified as unnecessary, but it will be a false positive in some CI pipeline on some random PR in the future.

Reply:
I think mypy is actually correct here and our typing for run() is wrong. The typing for run() should be `Sequence[object]` since it accepts anything that we can call `str()` on.

For proper typing we should probably turn that from `Sequence[object]` to `Sequence[str]`. It'll be kinda verbose to call `str()` on every path but it'll be safer overall since the type checker can pass when we pass something that isn't a string.

Reply:
Regardless, let's fix this in a separate PR since we do this in more than one place